evm/v0.5/contracts/dev/PrepaidAggregator.sol

```diff
@@ -0,0 +1,437 @@
+pragma solidity 0.5.0;
+
+import "../Median.sol";
+import "../vendor/Ownable.sol";
+import "../vendor/SafeMath.sol";
+import "./SafeMath128.sol";
+import "../interfaces/LinkTokenInterface.sol";
+import "../interfaces/WithdrawalInterface.sol";
+
+/**
+ * @title The Prepaid Aggregator contract
+ * @notice Node handles aggregating data pushed in from off-chain, and unlocks
+ * payment for oracles as they report. Oracles' submissions are gathered in
+ * rounds, with each round aggregating the submissions for each oracle into a
+ * single answer. The latest aggregated answer is exposed as well as historical
+ * answers and their updated at timestamp.
+ */
+contract PrepaidAggregator is Ownable, WithdrawalInterface {
+  using SafeMath for uint256;
+  using SafeMath128 for uint128;
+
+  struct Round {
+    int256 answer;
+    uint256 updatedTimestamp;
+    RoundDetails details;
+  }
+
+  struct RoundDetails {
+    int256[] answers;
+    uint64 maxAnswers;
+    uint64 minAnswers;
+    uint128 paymentAmount;
+  }
+
+  struct OracleStatus {
+    uint128 withdrawable;
+    uint128 startingRound;
+    uint128 endingRound;
+    uint128 lastReportedRound;
+    uint128 lastStartedRound;
+    int256 latestAnswer;
+  }
+
+  uint128 public latestRound;
+  uint128 public currentRound;
+  uint128 public allocatedFunds;
+  uint128 public availableFunds;
+
+  // Round related params
+  uint128 public paymentAmount;
+  uint64 public oracleCount;
+  uint64 public maxAnswerCount;
+  uint64 public minAnswerCount;
+  uint64 public restartDelay;
+
+  LinkTokenInterface private LINK;
+  mapping(address => OracleStatus) private oracles;
+  mapping(uint128 => Round) private rounds;
+
+  event NewRound(uint128 indexed number, address indexed startedBy);
+  event AnswerUpdated(int256 indexed current, uint128 indexed round);
+  event AvailableFundsUpdated(uint256 indexed amount);
+  event RoundDetailsUpdated(
+    uint128 indexed paymentAmount,
+    uint64 indexed minAnswerCount,
+    uint64 indexed maxAnswerCount,
+    uint64 restartDelay
+  );
+  event OracleAdded(address indexed oracle);
+  event OracleRemoved(address indexed oracle);
+
+  /**
+   * @notice Deploy with the address of the LINK token and initial payment amount
+   * @dev Sets the LinkToken address and amount of LINK paid
+   * @param _link The address of the LINK token
+   * @param _paymentAmount The amount paid of LINK paid to each oracle per response
+   */
+  constructor(address _link, uint128 _paymentAmount) public {
+    LINK = LinkTokenInterface(_link);
+    updateFutureRounds(_paymentAmount, 0, 0, 0);
```

Comment:
It was to emit an event and use one code path to avoid code duplication, but the event doesn't seem particularly necessary as it's unlikely that anyone will be subscribed to this address before the contract is deployed. Updated.